hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitionsAsync.java

```diff
@@ -0,0 +1,208 @@
+    /**
+     * Executes all the operations on the partitions.
+     */
+    @SuppressWarnings("unchecked")
+    <T> ICompletableFuture<Map<Integer, T>> invokeAsync(ExecutionCallback<Map<Integer, T>> callback) {
+        this.callback = (ExecutionCallback<Map<Integer, Object>>) ((ExecutionCallback) callback);
+        ensureNotCallingFromPartitionOperationThread();
+        invokeOnAllPartitions();
+        return future;
+    }
+
+    private void ensureNotCallingFromPartitionOperationThread() {
+        // TODO [viliam] is this check necessary when it's async?
+        if (Thread.currentThread() instanceof PartitionOperationThread) {
+            throw new IllegalThreadStateException(Thread.currentThread() + " cannot make invocation on multiple partitions!");
+        }
+    }
+
+    private void invokeOnAllPartitions() {
+        for (final Map.Entry<Address, List<Integer>> mp : memberPartitions.entrySet()) {
+            final Address address = mp.getKey();
+            List<Integer> partitions = mp.getValue();
+            PartitionIteratingOperation op = new PartitionIteratingOperation(operationFactory, toIntArray(partitions));
+            operationService.createInvocationBuilder(serviceName, op, address)
+                    .setTryCount(TRY_COUNT)
+                    .setTryPauseMillis(TRY_PAUSE_MILLIS)
+                    .invoke()
+                    .andThen(new FirstAttemptExecutionCallback(mp.getValue()));
+        }
+    }
+
+    private void retryPartition(final int partitionId) {
+        Operation operation;
+        PartitionAwareOperationFactory partitionAwareFactory = extractPartitionAware(operationFactory);
+        if (partitionAwareFactory != null) {
+            operation = partitionAwareFactory.createPartitionOperation(partitionId);
+        } else {
+            operation = operationFactory.createOperation();
+        }
+
+        operationService.createInvocationBuilder(serviceName, operation, partitionId)
+                        .invoke()
+                        .andThen(new ExecutionCallback<Object>() {
+                            @Override
+                            public void onResponse(Object response) {
+                                setPartitionResult(partitionId, response);
+                                decrementLatchAndHandle(1);
+                            }
+
+                            @Override
+                            public void onFailure(Throwable t) {
+                                setPartitionResult(partitionId, t);
+                                decrementLatchAndHandle(1);
+                            }
+                        });
+    }
+
+    private void decrementLatchAndHandle(int count) {
+        if (latch.addAndGet(-count) > 0) {
+            // we're not done yet
+            return;
+        }
+
+        Map<Integer, Object> result = createHashMap(partitionResults.length());
+        for (int partitionId = 0; partitionId < partitionResults.length(); partitionId++) {
```

Comment:
What if we invoked only on partitions `[5, 10, 5000]`?

Reply:
In this case we'll waste some memory, `partitionResults` is always allocated for all partitions. But this was an improvement from `Map<Integer, Object>` which used to be created with the capacity of `partitionCount`. Is this your point?

Reply:
👍